File: bindings/python/src/shell_use/_transport.py
from __future__ import annotations

import asyncio
import json
import os
from typing import Any, Dict, Optional, Tuple

from . import _config as cfg
from .errors import DaemonError

_Streams = Tuple[asyncio.StreamReader, asyncio.StreamWriter]


async def _open(session: str, home: Optional[str]) -> _Streams:
    path = cfg.socket_path(session, home)
    if cfg.IS_WINDOWS:
        loop = asyncio.get_running_loop()
        create = getattr(loop, "create_pipe_connection", None)
        if create is None:
            raise DaemonError(
                "named-pipe client requires the Proactor event loop on Windows "
                "(the default since Python 3.8)"
            )
        reader = asyncio.StreamReader()
        protocol = asyncio.StreamReaderProtocol(reader)
        transport, _ = await create(lambda: protocol, path)
        writer = asyncio.StreamWriter(transport, protocol, reader, loop)
        return reader, writer
    return await asyncio.open_unix_connection(path)


async def _close(writer: asyncio.StreamWriter) -> None:
    writer.close()
    try:
        await writer.wait_closed()
    except Exception:
        pass


async def can_connect(session: str, home: Optional[str]) -> bool:
    try:
        _, writer = await _open(session, home)
    except (FileNotFoundError, ConnectionRefusedError, OSError):
        return False
    await _close(writer)
    return True


async def ensure_daemon(session: str, home: Optional[str], binary: str) -> None:
    if await can_connect(session, home):
        return
    env = dict(os.environ)
    if home:
        env["SHELL_USE_HOME"] = home
    try:
        proc = await asyncio.create_subprocess_exec(
            binary,
            "--session",
            session,
            "daemon",
            "start",
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
            env=env,
        )
    except FileNotFoundError:
        raise DaemonError(
            f"could not find the '{binary}' binary on PATH; "
            "set SHELL_USE_BIN or pass binary="
        )
    await proc.wait()
    for _ in range(100):
        if await can_connect(session, home):
            return
        await asyncio.sleep(0.05)
    raise DaemonError(f"daemon for session '{session}' did not become ready")


async def request(
    session: str,
    home: Optional[str],
    binary: str,
    payload: Dict[str, Any],
    *,
    autostart: bool = True,
) -> Dict[str, Any]:
    if autostart:
        await ensure_daemon(session, home, binary)
    try:
        reader, writer = await _open(session, home)
    except (FileNotFoundError, ConnectionRefusedError, OSError) as e:
        raise DaemonError(f"could not connect to session '{session}': {e}")
    try:
        writer.write(json.dumps(payload).encode("utf-8") + b"\n")
        await writer.drain()
        line = await reader.readline()
    finally:
        await _close(writer)
    if not line:
        raise DaemonError("daemon closed the connection without responding")
    try:
        return json.loads(line.decode("utf-8"))
    except json.JSONDecodeError as e:
        raise DaemonError(f"invalid response from daemon: {e}")

File: bindings/js/src/transport.ts
import net from "node:net";
import { spawn } from "node:child_process";

import { socketPath } from "./config.js";
import { DaemonError } from "./errors.js";
import type { Response } from "./types.js";

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function spawnDaemon(
  binary: string,
  args: string[],
  env: NodeJS.ProcessEnv,
): Promise<void> {
  return new Promise((resolve, reject) => {
    const child = spawn(binary, args, { env, stdio: "ignore", windowsHide: true });
    child.once("error", reject);
    child.once("exit", () => resolve());
  });
}

function connect(target: string): Promise<net.Socket> {
  return new Promise((resolve, reject) => {
    const sock = new net.Socket();
    const onError = (err: Error) => {
      sock.destroy();
      reject(err);
    };
    sock.once("error", onError);
    sock.once("connect", () => {
      sock.removeListener("error", onError);
      resolve(sock);
    });
    sock.connect(target);
  });
}

export async function canConnect(session: string, home?: string): Promise<boolean> {
  try {
    const sock = await connect(socketPath(session, home));
    sock.destroy();
    return true;
  } catch {
    return false;
  }
}

export async function ensureDaemon(
  session: string,
  home: string | undefined,
  binary: string,
): Promise<void> {
  if (await canConnect(session, home)) {
    return;
  }
  const env = { ...process.env };
  if (home) {
    env.SHELL_USE_HOME = home;
  }
  try {
    await spawnDaemon(binary, ["--session", session, "daemon", "start"], env);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
      throw new DaemonError(
        `could not find the '${binary}' binary on PATH; set SHELL_USE_BIN or pass { binary }`,
      );
    }
    throw new DaemonError(`failed to start daemon: ${(err as Error).message}`);
  }
  for (let i = 0; i < 100; i++) {
    if (await canConnect(session, home)) {
      return;
    }
    await delay(50);
  }
  throw new DaemonError(`daemon for session '${session}' did not become ready`);
}

export async function request(
  session: string,
  home: string | undefined,
  binary: string,
  payload: unknown,
  autostart = true,
): Promise<Response> {
  if (autostart) {
    await ensureDaemon(session, home, binary);
  }
  let sock: net.Socket;
  try {
    sock = await connect(socketPath(session, home));
  } catch (err) {
    throw new DaemonError(
      `could not connect to session '${session}': ${(err as Error).message}`,
    );
  }
  return new Promise<Response>((resolve, reject) => {
    let buf = "";
    let settled = false;
    sock.setEncoding("utf8");
    sock.on("data", (chunk: string) => {
      buf += chunk;
      const nl = buf.indexOf("\n");
      if (nl >= 0 && !settled) {
        settled = true;
        sock.destroy();
        try {
          resolve(JSON.parse(buf.slice(0, nl)) as Response);
        } catch (err) {
          reject(new DaemonError(`invalid response from daemon: ${(err as Error).message}`));
        }
      }
    });
    sock.on("error", (err) => {
      if (!settled) {
        settled = true;
        reject(new DaemonError((err as Error).message));
      }
    });
    sock.on("close", () => {
      if (!settled) {
        settled = true;
        reject(new DaemonError("daemon closed the connection without responding"));
      }
    });
    sock.write(JSON.stringify(payload) + "\n");
  });
}
